src/rgw/rgw_req_context.h

```diff
@@ -0,0 +1,24 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#include "common/async/yield_context.h"
+namespace opentelemetry::trace {
+  class Span;
+}
+
+class DoutPrefixProvider;
+
+// this struct holds information which is created at the frontend
+// and should trickle down through all function calls to the backend
+struct req_context {
+  const DoutPrefixProvider* const dpp{nullptr};
```

Comment:
Do we want a way to stack dpp?  Or are we going to use a single dpp call down the stack?  I know the top is generally a RGWOp instance, but is this overridden in places?

Reply:
in general, it doesn't make much sense for a struct to have all-const members. if you want all-const, then your `req_context` variables should themselves be const. this could just a raw struct with all of the constructors/operators left defaulted

passing `req_context` around as a const-ref seems like the right approach. if a function wants to override parts of the context, it can construct a new variable with updated pointers:
```c++
void foo2(const req_context& rctx);

void foo(const req_context& rctx)
{
  // wrap the caller's dpp
  CustomDoutPrefix dpp{*rctx.dpp};
  // add a child span to the trace
  auto span = tracer.add_span("child", rctx.span);

  foo2({.dpp = &dpp, .y = rctx.y, .span = span.get()});
}
```

Reply:
> in general, it doesn't make much sense for a struct to have all-const members. 

according to your suggestion, in order to change anything with the context, you need to create a new context with thew new variables in it, and copy over what remains the same.
this means, that there are no changes to the struct once created. in this case, why not make them all const?

> if you want all-const, then your `req_context` variables should themselves be const. 

this is what i did, but making the data members const assures that even if the user does not make the `req_context` variable const

> this could just a raw struct with all of the constructors/operators left defaulted

why is this better?

Reply:
> according to your suggestion, in order to change anything with the context, you need to create a new context with thew new variables in it, and copy over what remains the same. this means, that there are no changes to the struct once created. in this case, why not make them all const?

because the const vs. mutable decision should be up to the caller instead of hard-coded into the type itself. what if a function wants to start a child span in the middle? there's nothing wrong with using a mutable context in a local scope:

```c++
void foo2(const req_context& rctx);
void foo3(const req_context& rctx);

void foo(const req_context& rctx)
{
  req_context rctx2{rctx}; // normal copy ctor
  foo2(rctx2);

  // start a child span just for foo3()
  auto span = tracer.add_span("child", rctx.span);
  rctx2.span = span.get();

  foo3(rctx2);
}
```

> this is what i did, but making the data members const assures that even if the user does not make the `req_context` variable const

the chosen convention is to pass by const-ref, because we don't want a called function to change the context of its parent function. this kind of const-correctness should already be part of our code reviews

> > this could just a raw struct with all of the constructors/operators left defaulted
> 
> why is this better?

a raw struct with normal value semantics is easier to reason about and less code to maintain. adding const to members disables operations like move and copy assignment. this can be useful for classes that manage a resource, or otherwise rely on encapsulation to maintain invariants. but `req_context` is just a bundle of related variables. let's keep it simple!